LRUCachingDAO

minsunny opened the pull request:
on foam
https://nanopay.atlassian.net/browse/CPF-311

https://github.com/nanoPayinc/NANOPAY/pull/6451

- initially for javascript client.
- only handle find, put, remove
- maintain an inner MapDAO as Cache
- on cache miss, call delegate (ClientDAO)
- on put, remove cache entry
- on delegate find
- found: add to cache
- not found: add key:NULL to cache to indicate that value does not exist in delegate, return null.


LRUCahingDAO applied on balanceDAO, userDAO,  invoiceDAO first,
Changed on Services file.

minsunny committed:
temp commit for test

minsunny committed:
LRUchachingDAO

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/CPF-311/LRUCachingDAO_

minsunny committed:
service

minsunny committed:
added LRUCachingDAO

minsunny committed:
modified easyDAO

jlhughes commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+  listeners: [
+    /** Adds the put() item to the tracking dao, runs cleanup() to check the dao size. */
+    function onPut(obj) {
+      var self = this;
```
Is it possible for onPut and onRemove to call the class methods put and remove above?

jlhughes commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+              self.put(x, o)
+              return o;
+            } else {
+              return null;
```
It may be future work, but are we planning on storing a NULL flag for misses?

kgrgreer replied:
The ItemCache should contain a boolean to say if the object exists or not. It will happen a lot that the DAO will be used to try to find a value that doesn't exist, and we need to cache that it doesn't exist otherwise we will still be going to the network every time.

kgrgreer commented on src/foam/dao/CacheType.js:
```diff
@@ -0,0 +1,31 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
```
2019

kgrgreer commented on src/foam/dao/CacheType.js:
```diff
@@ -0,0 +1,31 @@
+    },
+    {
+      name: 'FULL',
+      documentation: 'Using Regular CacheDAO',
```
'Use full caching.'

kgrgreer commented on src/foam/dao/CacheType.js:
```diff
@@ -0,0 +1,31 @@
+  values: [
+    {
+      name: 'NONE',
+      documentation: 'NO CacheDAO used',
```
'No caching'

kgrgreer commented on src/foam/dao/CacheType.js:
```diff
@@ -0,0 +1,31 @@
+    },
+    {
+      name: 'LRU',
+      documentation: 'Using LRU CacheDAO',
```
Use a LRU (Least Recently Used) partial cache

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -406,7 +414,7 @@ return delegate;
 //           });
 //         }
-        if ( this.cache ) {
+        if ( this.cacheType == 'FULL') {
```
Insert space before closing ).

kgrgreer replied:
Should be == foam.dao.CacheType.FULL

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -415,6 +423,14 @@ return delegate;
             of: this.model});
         }
+
+        if ( this.cacheType == 'LRU' ) {
```
Should be == foam.dao.CacheType.LRU

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -465,7 +481,7 @@ return delegate;
         dao.syncRecordDAO = foam.dao.EasyDAO.create({
           of: dao.SyncRecord,
-          cache: true,
+          cacheType: 'FULL',
```
Should be foam.dao.CacheType.FULL

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
```
2019

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+      /** The maximum size to allow the target dao to be. */
+      class: 'Int',
+      name: 'maxSize',
+      value: 100
```
10000 might be a better default

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -415,6 +423,14 @@ return delegate;
+        if ( this.cacheType == 'LRU' ) {
+          this.mdao = this.MDAO.create({of: params.of});
+          dao = this.LRUCachingDAO.create({
+            dao : this.mdao,
```
I think we could specify a factory on the 'dao' property of LRUCachingDAO so it would be easier to use. Setting it to an MDAO with 'of' being the same as the 'of' of the delegate is what users will do 99% of the time, so we should make this default.

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+    },
+    {
+      class: 'foam.dao.DAOProperty',
+      name: 'dao'
```
A more informative name like 'cacheDAO' would make the code easier to understand.

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+      this.SUPER();
+      var self = this;
+
+      self.delegate.listen(self.QuickSink.create({
```
This works, but is an unusual way to implement a decorator. It would be more efficient to just decorate the put(), remove() and removeAll() methods.

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+    /** Calculates a timestamp to use in the tracking dao. */
+    function getTimestamp() {
+      // Just increment on each request.
+      return this.lastTimeUsed_++;
```
If this is just a seqNo (which is good) then why make it be a timestamp at all, rather than just a seqNo starting at 0?

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+        .skip(self.maxSize)
+        .select({
+          put: function(obj) {
+            self.dao.remove(obj);
```
You also need to remove LRUCacheItem from the trackingDAO.

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+      name: 'trackingDAO',
+      of: 'foam.dao.DAO',
+      factory: function() {
+        return this.MDAO.create({ of: this.LRUCacheItem });
```
Add an index on the timestamp/seqNo field so that its efficient when you do the ORDER_BY, otherwise this will be very slow.

kgrgreer replied:
ping

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+      return this.trackingDAO.find_(x, id).then(function(o) {
+        if ( o ) {
+          return o;
+        } else {
```
No need for 'else' if you've already returned on the line above.

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,154 @@
+
+      return this.trackingDAO.find_(x, id).then(function(o) {
+        if ( o ) {
+          return o;
```
At this point, you should also update the timestamp of the object in the itemCache.

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/CPF-311/LRUCachingDAO_

minsunny committed:
handled the comments

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/CPF-311/LRUCachingDAO_

minsunny committed:
Merge branch 'master' of github.com:foam-framework/foam2 into f/CPF-311/LRUCachingDAO_

lchanmann commented on src/foam/dao/EasyDAO.js:
```diff
@@ -170,12 +172,18 @@ return delegate;
+      class: 'foam.core.Enum',
+      of: 'foam.dao.CacheType',
+      name: 'cacheType',
+      value: 'NONE' /* 'None'*/
```
Should it be
```suggestion
      value: foam.dao.CacheType.NONE /* 'None'*/
```

kgrgreer replied:
Both work because the adapt method for enums will make the conversion.
@mcarcaso Is it better to just do 'NONE' for class loading purposes?

minsunny replied:
Then will leave it as "NONE"

lchanmann commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,174 @@
+      /** By starting at 0 */
+      class: 'Int',
+      name: 'lastTimeUsed_',
+      factory: function() { return 0; }
```
can we use value instead of factory?

lchanmann commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,174 @@
+
+  classes: [
+    {
+      /** Links an object id to a last-accessed seqNo */
```
can we use documentation rather than code comment? There are also a few similar places below.

kgrgreer replied:
Yes, documentation: is better.

minsunny committed:
uptodate

minsunny committed:
fixed typo

minsunny committed:
fixed conflicts

minsunny committed:
added documentation

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -415,6 +424,13 @@ return delegate;
         }
+
+        if ( this.cacheType == foam.dao.CacheType.LRU ) {
+          this.mdao = this.MDAO.create({of: params.of});
```
This line isn't needed, is it? I don't think this.mdao is used by this cache.

kgrgreer replied:
ping

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,174 @@
+      name: 'LRUCacheItem',
+      properties: [
+        {
+          name: 'id',
```
No trailing comma.

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,174 @@
+    },
+    {
+      class: 'Int',
+      name: 'lastTimeUsed_',
```
The name 'lastTimeUsed_' is misleading since it isn't a time. Maybe something like 'nextSeqNo' would be better.

kgrgreer commented on src/foam/dao/CacheType.js:
```diff
@@ -0,0 +1,31 @@
+    {
+      name: 'LRU',
+      documentation: 'Using a LRU (Least Recently Used) partial cache',
+      label: 'LRU'
```
Label defaults to the same value as 'name' if not specified, so this line isn't needed.

kgrgreer commented on src/foam/dao/LRUCachingDAO.js:
```diff
@@ -0,0 +1,174 @@
+
+  documentation: `
+     Least Recently Used CachingDAO :
+     Manages a DAO\'s size by removing old items. Commonly applied inside a cache to limit the cache\'s size.
```
I don't think you need \' since this string isn't quoted with 's but is quoted with `.

minsunny committed:
handled commments

minsunny committed:
added documentation

minsunny committed:
added index

minsunny committed:
fixed conflicts on services
Merge branch 'master' of github.com:foam-framework/foam2 into f/CPF-311/LRUCachingDAO_

jlhughes commented:
Closed - replaced by javascript side caching.
